### DISCUSSION TITLE
CocHasProvider for non current buffer

### DISCUSSION THREAD

**Discussion by uga-rosa** [neoclide/coc.nvim#4664] [Q&A] [answered]
`CocHasProvider` targets the LS attached to the **current** buffer. I would like to do the same thing by specifying bufNr or langserver id. Is this possible? If you can get the entire server capacities from langserver id, that's fine.

**Comment by fannheyward** [comment]
> by specifying bufNr

No such feature yet.

> get the entire server capacities from langserver id

`client.initializeResult.capabilities` returns the server's capabilities. You can get client by `services.getService('test').client`.

**Reply by uga-rosa** [reply]
Thanks for the reply.

> client.initializeResult.capabilities returns the server's capabilities. You can get client by services.getService('test').client.

Please elaborate a bit more, I couldn't find anything like that in help. Is this a possible method from the vim script?

**Reply by fannheyward** [reply]
You can use `services.getService` in extensions, can't from vimscript.

What you most want is to check whether LS support X feature (by hasProvider) for current or special buffer?

**Reply by uga-rosa** [reply]
I am now using languageIds from the result of CocAction('services') to narrow down the LSs that should be valid on a certain buffer and send a CocRequest to them. However, since I cannot check the provider, I can only determine that the method is not supported if I detect an error in try/catch.

**Reply by fannheyward** [reply]
How about `hasProvider(id, bufnr?)`? When `bufnr` is missing, use current buffer.

**Reply by uga-rosa** [reply]
I'm sorry, I just want to know how it is possible from the vim script. `hasProvier()` is how I can access it?

**Reply by fannheyward** [reply]
`call CocAction('hasProvider', 'inlayHint')`

**Reply by uga-rosa** [reply]
Can I pass {id} or {bufnr} that way?

**Reply by fannheyward** [reply]
Not yet, `hasProvider` can only accept one param by now. I'll add `bufnr` later.

**Reply by fannheyward** [reply]
https://github.com/neoclide/coc.nvim/pull/4714

**Reply by uga-rosa** [reply]
Great work! Ty!

**Reply by fannheyward** [reply]
Sorry for the delay, I just missed this thread 😅

**Reply by fannheyward** [reply, marked as answer]
Use `hasProvider('rename', 1)`, 1 is the buffer number.
